### ISSUE TITLE
[Improvement] Scheduling using cost function.

### ISSUE THREAD

**Issue by SamMousa** [rancher/rancher#7886]
This post is about scheduling service on hosts based on the current available configuratoin only.
Thinks Rancher take into account:
- Number of containers (this is a simple utilization metric)
- Strict scheduling rules (must / must not)
- Preferred scheduling rules (should / should not).

I'd like to talk about the way Rancher chooses what host a new service should be deployed to.
First consider the set of all host `H`. From this set all hosts that are explicitly forbidden can be removed using the strict scheduling rules. This leaves us with a smaller set `H'`.
Since all strict rules have been taken into account, we can conclude a solution exists if `H'` is not empty.

This is the easy part, now come the part where Rancher has to decide how to divide the containers across the available hosts. Currently it seems Rancher only does this when initially launching a service and it is unclear when these conditions are re-evaluated.

Consider the galera cluster from the gallery, it comes with 3 nodes which, for obvious reasons, prefer to be scheduled on different hosts. When launching the cluster on a Rancher setup with 3 nodes all is fine.
If 2 hosts go down Rancher has no choice but to move all containers to the 1 remaining host.
Now when the hosts come back up (for simplicity lets assume it has been rebooted so the galera containers are not running anymore).
Now we're stuck in a situation where Rancher will not allow us to re-balance the galera cluster to make use of all hosts.
1. Trying to delete a container (and its sidekicks) from the host with all container results in the container being recreated on the same host.
2. Increasing the scale uses the proper hosts, but decreasing it afterwards then seems to remove the most recently created container.

All of these issues could be simply resolved if Rancher uses a cost function to compare different solution.
A simple cost function would be counting the number of broken non-strict rules. When a container needs to be removed there's only very few choices and it would be easy to calculate the best choice, same for adding a new container.

The only issue is initial scheduling where dividing `n` containers over `m` nodes gives us a really large search set, still improving this only for the rescheduling case would already be great.

tldr; In case Rancher needs to remove 1 container, could it do it intelligently instead of just removing the most recently created?

**Comment by superseb** [comment]
With the release of Rancher 2.0, development on v1.6 is only limited to critical bug fixes and security patches.
